### DISCUSSION TITLE
Preview feature feedback: `@prisma/adapter-pg` / PostgreSQL Driver support

### DISCUSSION THREAD

**Discussion by nikolasburk** [prisma/prisma#22899] [Feedback]
Support for the [`node-postgres`](https://node-postgres.com/) driver (`pg`) is in preview behind the `driverAdapters` preview feature.

Please share your feedback about `@prisma/adapter-pg` in this discussion.

- If you encounter a bug, please [open a bug report](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=bug_report.md&title=) in this repo.
- If the feature is working well for you, please share this in a comment below or leave a 👍 on this issue.
- If you have any questions, don't hesitate to ask them in the [Prisma Discord](https://pris.ly/discord) community

**Comment by theguacamoleking** [comment]
Hey @nikolasburk! 

The feature looks great but it seems that Prisma prepends all select statements with the `public` schema. 

```sql
SELECT "public".* FROM "TABLE"
```

Is there a way to get prisma to not include the schema in the resulting query — that way the user can [configure the default schema on the pool](https://stackoverflow.com/questions/15847326/is-it-possible-to-set-postgres-search-path-in-node-postgres)?

Alternatively, is there a way to inform Prisma of the schema?

**Reply by nikolasburk** [reply]
Ha, good question! Did you try configuring the schema on your connection string as documented [here](https://www.prisma.io/docs/orm/overview/databases/postgresql#arguments)? For example, assume your schema is called `myschema`:

```
postgresql://user:password@host:port/db?schema=myschema
```

Can you confirm that you did this and Prisma Client still prepends statements with the `public` schema name? If so, that's probably a bug that we should fix!

**Reply by SevInf** [reply]
Hi @theguacamoleking!
Configuring schema is done slightly differently when you are using adapter. If you want to use custom schema as default, use following code to instantiate your adapter:

```ts
const adapter = new PrismaPg(pool, {
  schema: 'myDbSchema'
})
```

This is something the slipped into release without documentation, I've created [an issue](https://github.com/prisma/docs/issues/5722) to fix this.

**Reply by rstreefland** [reply]
@SevInf Just a heads up that a custom schema configured when instantiated using the adapter doesn't work for `$queryRaw` queries. [I've raised a bug for this](https://github.com/prisma/prisma/issues/24660) 😃

**Reply by theguacamoleking** [reply]
Similarly, it looks like schema doesn't work with `$queryRawTyped` (TypedSQL) - https://github.com/prisma/prisma/issues/26930

**Reply by theguacamoleking** [reply]
@nikolasburk there's a proposed fix here - https://github.com/prisma/prisma/pull/26935

**Comment by alonbilu** [comment]
Hi
Not sure that's the right place to ask, but.. :)
Anyways, I read the docs but couldn't find an answer for why would one consider switching to Node.js's native driver over using the one built in with Prisma...minding I'm connecting to PostgreSQL server, are there any benefits of doing this transition?
Thanks for clarifying this.

**Reply by janpio** [reply]
Good question! 

In certain environments, the built-in drivers are not available - for example on Cloudflare Workers or Vercel Edge Functions because they run a non-Node runtime that does not support the kind of functionality required (TCP socket support). For some of these cases, you can use `@prisma/adapter-pg` instead where support - Cloudflare makes that possible with their `connect()` TCP feature which was implemented by `pg` - so you can talk to your normal PostgreSQL database from Cloudflare Workers using `pg` via `@prisma/adapter-pg`.

Another angle might be that you just prefer the APIs of `pg` for configuration or connection pooling. With `@prisma/adapter-pg` you _can_ do that if you want.

(If you have no pain point that makes you move away from the built-in drivers, then you probably don't need to.)

**Reply by srujun** [reply]
Is it expected that using `@prisma/adapter-pg` speeds up Prisma operations significantly as compared to using the native Prisma driver? I am running a test in which I am sending around 40 queries concurrently, and repeating this a few times. I am noticing that the first few queries are much slower when using the Prisma driver and take around 200ms-300ms to complete, and then subsequent queries are able to finish within 10ms-20ms. When using `@prisma/adapter-pg`, the first few queries all finish within 20ms-50ms, and subsequent queries are able to finish within the same 10ms-20ms duration. In both cases I am using a connection pool of size 20 (using `connection_limit=20` in the connection string when using the native Prisma driver, and setting `max: 20` in `pg.Pool` when using `@prisma/adapter-pg`).

Why is the native Prisma driver slower?

**Reply by janpio** [reply]
That can have all kinds of reasons, srujun. The native driver for example has a more secure default configuration than `pg`, which with some database configurations makes opening connections slower. If you can reproduce this, please open an issue with reproduction steps and we can take a look.

**Reply by peterHakio** [reply]
I guess it is linked to https://github.com/prisma/prisma/issues/23573

**Comment by rstreefland** [comment]
I just wanted to highlight that this feature is incredibly useful for us because it makes using Prisma with passwordless DB authentication possible [despite it not currently being supported out of the box](https://github.com/prisma/prisma/issues/7869).

For example using IAM passwordless login via CloudSQL proxy for Google Cloud SQL using Cloud Run or Cloud Functions:

```typescript
import { PrismaPg } from "@prisma/adapter-pg";
import { PrismaClient } from "@prisma/client";
import { GoogleAuth } from "google-auth-library";
import { Pool } from "pg";

const auth = new GoogleAuth({
  scopes: ["https://www.googleapis.com/auth/sqlservice.login"],
});

const pool = new Pool({
  host: "/cloudsql/REDACTED",
  user: "REDACTED",
  password: async (): Promise<string> => {
    return (await auth.getAccessToken()) as string;
  },
  database: "postgres",
  ssl: false,
  idleTimeoutMillis: 0,
  max: 5,
});

const adapter = new PrismaPg(pool);

export const prismaClient = new PrismaClient({
  adapter,
});
```

I hope it reaches general availability soon 🤞

**Reply by ziazon** [reply]
Curious, how did you get migrations to work?  I have a config file like this for my app run and seed script, but I can't seem to find a way to have the migration command use this connection config file.

**Comment by Blitz2145** [comment]
Thanks for adding the pg driver adapter team! Would love to see driver adapters reach GA. We switched to the driver adapter in prod recently and its been working very well. Idle connections are actually closed properly compared to the native driver, since the native driver connection pool of mobc does not close idle conns after a timeout like most other pools (ex: https://github.com/brettwooldridge/HikariCP?tab=readme-ov-file#frequently-used and https://node-postgres.com/apis/pool#new-pool).

**Comment by theguacamoleking** [comment]
Thanks for the feature! Based on the implementation, it looks like [@prisma/extension-read-replicas](https://github.com/prisma/extension-read-replicas/tree/main) won't work in combination with this postgres driver. Are there any plans for supporting this in the future (maybe on the adapter itself)?

**Edit:** Looks like read replicas are [currently broken](https://github.com/prisma/extension-read-replicas/issues/12) even with the native DB driver.

**Comment by jasonhilldm** [comment]
I am trying to use this adapter with a Vike + Hono site deployed to Cloudflare Pages but receive the following error when trying to read from the database:

```
TypeError: Illegal invocation: function called with incorrect `this` reference. See https://developers.cloudflare.com/workers/observability/errors/#illegal-invocation-errors for details.
    at bundledWorker-0.6505292954156219.mjs:17468:23
    at async PrismaPg.performIO (bundledWorker-0.6505292954156219.mjs:18462:26)
    at async PrismaPg.queryRaw (bundledWorker-0.6505292954156219.mjs:18420:21)
    at async bundledWorker-0.6505292954156219.mjs:20151:18

```

Any pointers?

**Comment by beiserjohannes** [comment]
If before creating the Prisma Client I add my own type parser like so:

```
import { types as pgTypes } from 'pg';

pg.types.setTypeParser(1082, (val) => `DATEONLY-${val}`);

```

This should make Prisma convert Dates using this mapping function.
However, the following code from `prisma/adapter-pg/index.js` [line 107](https://github.com/prisma/prisma/blob/main/packages/adapter-pg/src/pg.ts) makes it so that if there is a custom-parser (internally from prisma) that one is used instad of my overwrite.
Is this the desired behaviour? It doesn't allow custom parsing like in my case differentiating treatment of Date and DateTime fields
```
 getTypeParser: (oid: number, format?) => {
      if (format === 'text' && customParsers[oid]) {
        return customParsers[oid]; //<----- here the prisma internal Date mapper is used instead of my global override
      }

      return types.getTypeParser(oid, format); // logging this does in fact include my custom mapper
 }

```

**Comment by aifirstd3v** [comment]
This is my `db.ts` for Prisma with pool with NextJS 15.x and Prisma 6.x.

`new PrismaPg` doesn't accept the 'pool' instance as an argument but `connectionString` object.


`postgresql://USER:PASSWORD@HOST:PORT/DATABASE?schema=myschema&connection_limit=5&socket_timeout=3`

```ts
/**
 * Prisma Client initialization with Singleton pattern.
 * This module exports a single Prisma Client instance named 'db' for use across your application.
 *
 * Usage:
 * ```typescript
 * import db from 'path_to_this_file';
 *
 * Example usage
 * const users = await db.user.findMany();
 * ```
 */

import { Prisma, PrismaClient } from '@prisma/client'
import { Pool } from 'pg'
import { PrismaPg } from '@prisma/adapter-pg'
//import { PrismaPg } from '@prisma/adapter-pg-worker'

// Extend the global object to include 'prisma' property
declare global {
  // eslint-disable-next-line no-var
  var prisma: PrismaClient | undefined
}

// Initialize the PostgreSQL connection pool
// const pool = new Pool({ connectionString: process.env.DATABASE_URL })

// Initialize the PrismaPg adapter with the connection pool, The PrismaPg adapter is the component that connects Prisma to the PostgreSQL database using the specified configuration for connection management.
const adapter: PrismaPg = new PrismaPg(
  { connectionString: process.env.DATABASE_URL },
  {
    // Optional: specify schema if needed
    // schema: 'myPostgresSchema',
  }
)

// Extend PrismaClient to accept the custom adapter
interface CustomPrismaClientOptions extends Omit<Prisma.PrismaClientOptions, 'datasources'> {
  adapter: PrismaPg
}

class CustomPrismaClient extends PrismaClient {
  constructor(options: CustomPrismaClientOptions) {
    super(options as Prisma.PrismaClientOptions) // Cast to base type
  }
}

// Access the global object and extend it with 'prisma' if necessary
const globalForPrisma = global as typeof global & { prisma?: PrismaClient }

// Create a singleton instance of PrismaClient
export const db =
  globalForPrisma.prisma ||
  new CustomPrismaClient({
    adapter,
    log: [
      {
        emit: 'event',
        level: 'query'
      },
      'info',
      'warn',
      'error'
    ]
  })

// Attach an event listener for query events (optional)
db.$on('query' as never, async (e: Prisma.QueryEvent) => {
  // Define a list of tables to exclude from logging
  const excludedTables = ['"lms"."User"', '"lms"."Account"'];

  // Check if the query is for any of the excluded tables
  if (excludedTables.some(table => e.query.includes(`FROM ${table}`))) {
    return; // Skip logging for excluded tables
  }

  // Log the query if it does not match the excluded tables
  console.log('Query: ' + e.query);
  console.log('Params: ' + e.params);
  console.log('Duration: ' + e.duration + 'ms');
});


// In development mode, store the instance globally to prevent multiple instances
if (process.env.NODE_ENV !== 'production') {
  globalForPrisma.prisma = db
}

// Export the PrismaClient instance as the default export
// export default db
```

**Reply by hyusetiawan** [reply]
im facing the same issue, it seems we do not need pg anymore and pass the connection settings directly to prismapg adapter? does this work for cloudflare worker environment as well?

**Comment by marlon360** [comment]
I am facing the problem that enum arrays are not parsed correctly when using `$queryRaw`. 

I am receiving `"{VALUE_1,VALUE2}"` instead of `["VALUE_1", "VALUE_2"]`

**Comment by fatihyildiz-cs** [comment]
Hello, any suggestions for the fix of this issue?
https://github.com/prisma/prisma/discussions/27496

**Comment by GauthierPLM** [comment]
I recently switched to pg adapter + prisma 6.16 and noticed that since then, the tables and types in migrations are prefixed with the schema name.

In our E2E tests, we pop a test container and use one schema per test suite. Because the schema name is present in the migrations, the DB setup fails (prisma complains that identifiers prefixed with `public` as not found, since they are in the current schema and not `public`).

The temporary solution is to clean manually the migrations before merging PRs, but this is time consuming and error-prone (already pushed 3 bad migrations while working on a new feature).

Is there any way to restore the previous behaviour which omitted the schema name in migrations?

**Comment by shoxter** [comment]
Recently switched to the adapter -- any reason we're hard-coding `false` on using booleans for `tinyint(1)`? Seems like this should be a developer choice.
